### DISCUSSION TITLE
Extended MMX (Cyrix EMMI) instructions

### DISCUSSION THREAD

**Discussion by Torinde** [bochs-emu/Bochs#302] [Ideas]
What do you think of emulating [Extended MMX (Cyrix EMMI)](https://en.wikipedia.org/wiki/List_of_discontinued_x86_instructions#Cyrix_EMMI_instructions) instructions?

- There is a [CPU flag in Linux](https://unix.stackexchange.com/questions/43539/what-do-the-flags-in-proc-cpuinfo-mean) for those: `cxmmx` Cyrix MMX extensions
- [Cyrix Extension to the Multimedia Instruction Set, Application Note 108](http://hackipedia.org/browse.cgi/Hardware/CPU/x86/Cyrix%206x86/Cyrix%20Extension%20to%20the%20Multimedia%20Instruction%20Set%2c%20Application%20Note%20108%2c%20Applies%20to%20the%20Cyrix%20MII%e2%84%a2%20CPU%20%281998-09-09%29.pdf) (some further docs and software are linked [here](https://github.com/PCBox/PCBox/issues/39#issuecomment-1147292376))
- Opcodes overlap with SSE and other subsequent instructions
- https://github.com/86Box/86Box/pull/5284

**Comment by stlintel** [comment]
Doesn't seem complicated. But do you have any program that uses them (for testing) ?
And which CPU model they will be enabled ?
I could create Cyrix model for fun - but what else is missing these except these instructions ?

**Reply by Torinde** [reply]
> program for testing

I've found only [some related CPU utilities](https://github.com/PCBox/PCBox/issues/39#issuecomment-1147292376)

> which CPU model they will be enabled ?

Cyrix 6x86MX, MediaGXm, Geode GXm/GXLV/GX1. Not sure about GX2/LX.

> what else is missing these except these instructions ?

A few System Management mode (and iGPU-related for MediaGX/Geode) [instructions](https://en.wikipedia.org/wiki/List_of_discontinued_x86_instructions#Instructions_specific_to_Cyrix_and_Geode_CPUs)
A few [x87 instructions](https://en.wikipedia.org/wiki/List_of_discontinued_x86_instructions#Instructions_present_in_specific_80387_clones)
A few [3DNow! instructions](https://en.wikipedia.org/wiki/List_of_discontinued_x86_instructions#3DNow!_instructions_specific_to_Geode_GX_and_LX) (for GX2/LX)

**Comment by stlintel** [comment]
Again - do we really need them ?
For example because we know how to build a CPUID model which will enable them. And do we have a set of programs which can be run with this new model and it would be nice to have them running ?

Note even most of the 3DNOW is PANIC(not implemented) in Bochs right now.

**Reply by Torinde** [reply]
Besides the CPU config utilities I haven't found other software.
As for CPUID model - there are [CPUID Dumps](http://users.atw.hu/instlatx64/CyrixInstead/CyrixInstead0000600_MII_CPUID.txt) and I assume we can find datasheets for both 6x86MX and MediaGXm/Geode.

I agree 3DNow! is higher priority and the Cyrix part is more about fun, than need. :)

**Reply by Torinde** [reply]
6x86MX/MII, 300 MHz (3 x 100) PR433 [CPUID dump](http://users.atw.hu/instlatx64/CyrixInstead/CyrixInstead0000601_MII_CPUID.txt)
MediaGXm, 233 MHz (7 x 33) [CPUID dump](http://users.atw.hu/instlatx64/CyrixInstead/CyrixInstead0000540_MediaGXm_CPUID.txt)

Documentation:
- [Cyrix 6x86MX](http://hackipedia.org/browse.cgi/Hardware/CPU/x86/Cyrix%206x86), [IBM 6x86MX](http://hackipedia.org/browse.cgi/Hardware/CPU/x86/IBM%206x86/MX), [IBM 6x86 (non-MX)](http://hackipedia.org/browse.cgi/Hardware/CPU/x86/IBM%206x86)
- [Cyrix MII](http://hackipedia.org/browse.cgi/Hardware/CPU/x86/Cyrix%20M%20II)

**Reply by luibgn** [reply]
No. We don't need them.
I think 3DNow! is fine after the improvements in the last months. Is there something still missing in 3DNow! ?

**Reply by Torinde** [reply]
> Is there something still missing in 3DNow! ?

Maybe the two Geode GX/LX instructions:
- PFRSQRTV – Reciprocal square root approximation for a pair of 32-bit floats
- PFRCPV – Reciprocal approximation for a pair of 32-bit floats
